gen/map.go

```diff
@@ -190,6 +190,111 @@ func (m *mapGenerator) Reader(g Generator, spec *compile.MapSpec) (string, error
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
 
+// Equals generates a function to compare maps of the given type
+//
+// 	func $name(lhs, rhs $mapType) bool {
+// 		...
+// 	}
+//
+// And returns its name.
+func (m *mapGenerator) Equals(g Generator, spec *compile.MapSpec) (string, error) {
+	if !isHashable(spec.KeySpec) {
+		return m.equalsUnhashable(g, spec)
+	}
+
+	name := "_" + valueName(spec) + "_EqualsHashable"
+	err := g.EnsureDeclared(
+		`
+			<$mapType := typeReference .Spec>
+
+			<$lhs := newVar "lhs">
+			<$rhs := newVar "rhs">
+			func <.Name>(<$lhs>, <$rhs> <$mapType>) bool {
+				if len(<$lhs>) != len(<$rhs>) {
+					return false
+				}
+
+				<$lk := newVar "lk">
+				<$lv := newVar "lv">
+				<$rv := newVar "rv">
+				<$ok := newVar "ok">
+				for <$lk>, <$lv> := range <$lhs> {
+					<$rv>, <$ok> := <$rhs>[<$lk>]
+					if !<$ok> {
+						return false
+					}
+					if !<equals .Spec.ValueSpec $lv $rv> {
+						return false
+					}
+				}
+				return true
+			}
+		`,
+		struct {
+			Name string
+			Spec *compile.MapSpec
+		}{Name: name, Spec: spec},
+	)
+
+	return name, wrapGenerateError(spec.ThriftName(), err)
+}
+
+func (m *mapGenerator) equalsUnhashable(g Generator, spec *compile.MapSpec) (string, error) {
+	name := "_" + valueName(spec) + "_EqualsUnhashable"
+
+	err := g.EnsureDeclared(
+		`
+			<$mapType := typeReference .Spec>
+
+			<$lhs := newVar "lhs">
+			<$rhs := newVar "rhs">
+			func <.Name>(<$lhs>, <$rhs> <$mapType>) bool {
+				if len(<$lhs>) != len(<$rhs>) {
+					return false
+				}
+
+				<$i := newVar "i">
+				<$j := newVar "j">
+				<$lk := newVar "lk">
+				<$lv := newVar "lv">
+				<$rk := newVar "rk">
+				<$rv := newVar "rv">
+				<$ok := newVar "ok">
+				for _, <$i> := range <$lhs> {
```

Comment:
This code is O(n^2). It might be possible to go O(n log n) by sorting lhs and rhs before lopping over both at the same time.